### PR TITLE
Refactor integration and `QuadraticInterpolation`

### PR DESCRIPTION
**Pull request by SouthEndMusic**
@sathvikbhagavan I split this out from https://github.com/SciML/DataInterpolations.jl/pull/356 for ease of reviewing. I need the `QuadraticInteprolation` refactor for the integration refactor, and the integration refactor for https://github.com/SciML/DataInterpolations.jl/pull/356.

### CODE REVIEW

File: ext/DataInterpolationsChainRulesCoreExt.jl [hunk 1]
```diff
@@ -74,6 +74,11 @@ function u_tangent(A::LinearInterpolation, t, Δ)
     out
 end
 
+function _quad_interp_indices(A::QuadraticInterpolation, t::Number, iguess)
```
Comment (sathvikbhagavan) [inline ext/DataInterpolationsChainRulesCoreExt.jl:77]: Just to confirm, can't `u_tangent` for QuadraticInterpolation be reformulated using `α` and `β` parameters you added?

**Review by sathvikbhagavan — COMMENTED** [review]
Left one comment. Are we getting any speedup from this implementation of QuadraticInterpolation than the earlier version? Otherwise, Looks good!

File: ext/DataInterpolationsChainRulesCoreExt.jl (hunk 1 above)
Comment (SouthEndMusic) [reply ext/DataInterpolationsChainRulesCoreExt.jl:77]: I'ts possible, but I suspect this is more efficient. Also, that might break AD types that do not support mutation when those parameters are cached?

Comment (SouthEndMusic) [reply ext/DataInterpolationsChainRulesCoreExt.jl:77]: And you still need to have the logic somewhere to see which mode (forward/backward looking interpolation) was used for the `u` gradient.

Comment (sathvikbhagavan) [reply ext/DataInterpolationsChainRulesCoreExt.jl:77]: Alright, then we can keep this.

**Review by sathvikbhagavan — APPROVED** [review]

**Review by ChrisRackauckas — APPROVED** [review]

**Pull request merged**